chef/automate
# [A2-833] Generalize project introspection

### :nut_and_bolt: Description

IntrospectAllProjects was evaluating a subset of endpoints in determining the set of allowed projects--not taking into account _parameterized_ endpoints. It turns out that for determining projects we can safely ignore action/resource pairs, which makes the parameterized question moot to boot.

### :+1: Definition of Done

The global projects filter will populate with all allowed projects, irrespective of whether they are on parameterized endpoints or not.

![image](https://user-images.githubusercontent.com/6817500/58596193-fe1a9080-8227-11e9-9812-86f72cabee06.png)


### :athletic_shoe: Demo Script / Repro Steps

```
export TOK=`chef-automate admin-token`

chef-automate iam upgrade-to-v2 --beta2.1 --skip-policy-migration

# Create a project
jo id='proj3' name='Project 3' | \
curl -sSkH "api-token: $TOK" https://localhost/apis/iam/v2beta/projects --data @-  | jq -c

# Create a non-admin user
jo name=coolbean id=coolbean password=chefautomate | \
curl -sSkH "api-token: $TOK" https://localhost/apis/iam/v2beta/users --data @-  | jq -c

# save below as policy.json
{
  "id": "test",
  "name": "testing global project filter",
  "members": ["user:local:coolbean"],
  "projects": ["proj3"],
  "statements": [
    {
      "effect": "ALLOW",
      "actions": ["iam:teams:get"],
      "projects": ["proj3"]
    }
  ]
}
curl -sSkH "api-token: $TOK" -d @policy.json https://localhost/apis/iam/v2beta/policies | jq
```

(Above leverages sample repro steps on the jira card--thanks @bcmdarroch!)

With all that in place: 
- login as user `coolbean` and you will see `project-p1` as the one and only project in the global project filter.
- login as user `admin` and you will see that plus `(unassigned)` plus any others you happen to have in your system.

### :chains: Related Resources

### :white_check_mark: Checklist
